hathor/transaction/storage/migrations/migrate_static_metadata.py

```diff
@@ -0,0 +1,67 @@
+#  Copyright 2023 Hathor Labs
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from typing import TYPE_CHECKING
+
+from structlog import get_logger
+
+from hathor.conf.get_settings import get_global_settings
+from hathor.transaction import Block, Transaction
+from hathor.transaction.static_metadata import BlockStaticMetadata, TransactionStaticMetadata
+from hathor.transaction.storage.migrations import BaseMigration
+from hathor.util import progress
+
+if TYPE_CHECKING:
+    from hathor.transaction.storage import TransactionStorage
+
+logger = get_logger()
+
+
+class Migration(BaseMigration):
+    def skip_empty_db(self) -> bool:
+        return True
+
+    def get_db_name(self) -> str:
+        return 'migrate_static_metadata'
+
+    def run(self, storage: 'TransactionStorage') -> None:
+        """This migration takes attributes from existing vertex metadata and saves them as static metadata."""
+        log = logger.new()
+        settings = get_global_settings()
+
+        # First we migrate static metadata using the storage itself since it uses internal structures.
+        log.info('creating static metadata...')
+        storage.migrate_static_metadata(log)
+
+        # Now that static metadata is set, we can use the topological iterator normally
+        log.info('removing old metadata and validating...')
+        topological_iter = storage.topological_iterator()
+
+        for vertex in progress(topological_iter, log=log, total=None):
+            # We re-save the vertex's metadata so it's serialized with the new `to_bytes()` method, excluding fields
+            # that were migrated.
+            storage.save_transaction(vertex, only_metadata=True)
+
+            # We re-create the static metadata from scratch and compare it with the value that was created by the
+            # migration above, as a sanity check.
```

Comment:
Exactly. Ideally only one loop would be necessary, with the `topological_iterator`, and then I would simply recalculate everything from scratch. However I cannot use the topological iterator initially because I cannot retrieve transactions from the storage — the retrieval would fail for not having static metadata yet.

Hence, two loops are needed. The first one exists to initialize the static metadata. It could also simply set "fake" values such as `0`, but I decided that importing real values from existing metadata and then comparing them with the recalculation in the second loop would be more reliable.

In theory the topological iterator could exist without any metadata, so maybe that's a sign that we should refactor this mechanism. But that could be done in a separate PR.

Reply:
@glevco There is a topological iterator that doesn't need any metadata (`_topological_sort_dfs`), but it is significantly slower.

Reply:
@jansegre the issue is not with the iterator itself, but internally it uses methods to get vertices from the storage, and those methods expect that metadata/static metadata are set.

Reply:
> the issue is not with the iterator itself, but internally it uses methods to get vertices from the storage, and those methods expect that metadata/static metadata are set.

It's not too clear to me. But the current approach is the best IMO, it's probably much faster to do two faster loops than one slow loop of `_topological_sort_dfs`.

Reply:
Sure, I'll leave it like this for this PR.

But just to clarify, `_topological_sort_dfs` ends up calling this method:

```python
    def _get_tx(self, hash_bytes: bytes, tx_data: bytes) -> 'BaseTransaction':
        tx = self.get_transaction_from_weakref(hash_bytes)
        if tx is None:
            meta_data = self._db.get((self._cf_meta, hash_bytes))
            tx = self._load_from_bytes(tx_data, meta_data)
            self._load_static_metadata(tx)
            assert tx.hash == hash_bytes
            self._save_to_weakref(tx)
        return tx
```

And therefore it requires that both metadata and static metadata are set and can be correctly deserialized.